### ISSUE TITLE
Now we can do as('a').out.as('b') in Gremlin-Groovy with SugarPlugin.

### ISSUE THREAD

**Issue by okram** [tinkerpop/tinkerpop3#463]
@mbroecheler @joshsh @dkuppitz 

Now that we have `AnonymousGraphTraversal.Tokens.__`, `match()` looks like this in Gremlin-Groovy (with Sugar).

```groovy
g.V.match('a',
    __.as('a').out('created').as('b'),
    __.as('b').has('name', 'lop'),
    __.as('b').in('created').as('c'),
    __.as('c').has('age', 29)).
                 select('a', 'c').by('name')
```

The only difference between this and non-sugar is `g.V` vs. `g.V()`. However, if we add to SugarPlugin that a missing method named `as()` is actually a called to `AnonymousGraphTraversal.Tokens.__` (which is 3 lines of code :)), then it will look like this:

```groovy
g.V.match('a',
    as('a').out('created').as('b'),
    as('b').has('name', 'lop'),
    as('b').in('created').as('c'),
    as('c').has('age', 29)).
                 select('a', 'c').by('name')
```

This is cool, but realize it only works for Gremlin-Groovy (with Sugar). Be nice to come up with a simplification that would also work for non-sugar and Gremlin-Java.

**Comment by mbroecheler** [comment]
Yeah, that is nice. Crazy idea: would it be possible to extend the sugar such that uppercase words are treated as "as" declarations. That would look really pretty in match:

```groovy
g.V.match('a',
    A.out('created').B,
    B.has('name', 'lop'),
    B.in('created').C,
    C.has('age', 29)).
                 select('A', 'C').by('name')

```

**Comment by okram** [comment]
You could, but think about it when its words, not just letters.

```groovy
g.V.match('person',
    PERSON.out('created').PROJECT,
    PROJECT.has('name', 'lop'),
    PROJECT.in('created').COLLABORATOR,
    COLLABORATOR.has('age', 29)).
                 select('person', 'collaborator').by('name')
```

looks sorta intense.... ? .... Instead of:

```groovy
g.V.match('person',
    as('person').out('created').as('project'),
    as('project').has('name', 'lop'),
    as('project').in('created').as('collaborator'),
    as('collaborator').has('age', 29)).
                 select('person', 'collaborator').by('name')
```

Next, camel-case terms? ... "workPlace"... Finally, it really starts to pull Gremlin-Groovy further from Gremlin-Java. 

You could do this, but still "pull further from Gremlin-Java."

```groovy
g.V.match('a',
    _a.out('created')._b,
    _b.has('name', 'lop'),
    _b.in('created')._c,
    _c.has('age', 29)).
                 select('a', 'c').by('name')

---and---

g.V.match('person',
    _person.out('created')._project,
    _project.has('name', 'lop'),
    _project.in('created')._collaborator,
    _collaborator.has('age', 29)).
                 select('person', 'collaborator').by('name')
```

That is, any language token that is unknown and starts with `_`  is assumed to be `__.as()`.

**Comment by mbroecheler** [comment]
I was thinking that the criterion would be "word that starts with an upper case letter" since variable and method names by convention are lower case:

```groovy
g.V.match('person',
    Person.out('created').Project,
    Project.has('name', 'lop'),
    Project.in('created').Collaborator,
    Collaborator.has('age', 29)).
                 select('person', 'collaborator').by('name')
```

That reads extremely well to me. Also works nicely with convention since Person, Project, etc are all classes of things which are written in camelCase. Now, that would clash with imported class names, but the likelihood of collision seems small (since users will mostly use single letter or short variable names) and easy to detect.

Next, I don't mind that the sugar is different from the pure Java form. In a way, that's what a sugar is for: provide syntactic relieve. However, it does not change the nature of the language. That's what I would be more concerned about. As long as the sugar is materially the same as the pure form, I think it will be easy for users to move between the variants as the grammar and step names (which is the hard part to learn) are identical.
